backend-architecture-api.puml

```diff
@@ -0,0 +1,178 @@
+@startuml
+
+class ApiController {
+    +Index index()
+    +ResponseEntity<LandscapeImpl> landscape(String landscapeIdentifier)
+    +ResponseEntity<Group> group(String landscapeIdentifier, String groupIdentifier)
+    +ResponseEntity<Item> item(String landscapeIdentifier, String groupIdentifier, String itemIdentifier)
+    +ProcessLog create(@RequestBody String body)
+    +ProcessLog indexLandscape(String identifier, String format, String body)
+    +ResponseEntity<ProcessLog> log(String identifier)
+    +ResponseEntity<Set<Item>> search(String identifier, String query)
+    +ResponseEntity<List<FacetResult>> facets(String identifier)
+    +ProcessLog reindex(String landscape)
+}
+
+ApiController o-- LandscapeRepository
+ApiController o-- LandscapeDescriptionFactory
+ApiController o-- ItemDescriptionFormatFactory
+ApiController o-- Indexer
+
+class LandscapeImpl {
+    -String identifier
+    -String name
+    -String contact
+    -String description
+    -String source
+    -ItemIndex items
+    -LandscapeConfig config
+    -Map<String, GroupItem> groups
+    -ProcessLog processLog
+    -Map<String, String> labels
+    -Map<String, Link> links
+    -String owner
+}
+
+class LandscapeConfig {
+    -boolean greedy = true;
+    -LayoutConfig groupLayoutConfig
+    -LayoutConfig itemLayoutConfig
+    -List<String> groupBlacklist
+    -List<String> labelBlacklist
+    -Branding branding
+    -Map<String, KPIConfig> kpis
+}
+
+class LayoutConfig {
+    -Integer maxIterations
+    -Float forceConstantFactor
+    -Float maxDistanceLimitFactor
+    -Float minDistanceLimitFactor
+}
+
+LandscapeImpl o-- LandscapeConfig
+LandscapeConfig o-- LayoutConfig
+
+class LandscapeRepository {
+    +Optional<LandscapeImpl> findDistinctByIdentifier(String identifier)
+    +void save(LandscapeImpl landscape)
+    +Iterable<LandscapeImpl> findAll()
+}
+
+LandscapeRepository o-- LandscapeImpl
+
+class LandscapeDescriptionFactory {
+    +LandscapeDescription from(Landscape landscape)
+    +LandscapeDescription from(URL url)
+    +LandscapeDescription fromYaml(File file)
+    +LandscapeDescription fromString(String yaml, String origin)
+    +LandscapeDescription fromString(String yaml, URL url)
+}
+
+LandscapeDescriptionFactory -- LandscapeDescription
+
+class LandscapeDescription{
+    -String identifier
+    -String name
+    -String contact
+    -String description
+    -String owner
+    -Map<String, ItemDescription> templates
+    -String source
+    -List<SourceReference> sources
+    -ItemDescriptions itemDescriptions
+    -LandscapeConfig config
+    -boolean isPartial
+    -Map<String, GroupItem> groups;
+    -Map<String, Link> links
+    -Map<String, String> labels
+
+}
+
+class ItemDescriptionFormatFactory {
+    +ItemDescriptionFactory getFactory()
+}
+
+class Indexer {
+Triggered by ApplicationEvents fired by observing changes in files through the
+FileSourceReferenceObserver
+--
+    -LandscapeRepository landscapeRepo
```

Comment:
Is this the only one who triggers the Indexer?

So the Observers are only for registering file changes, right? And then an ApplicationEvent is fired, which in turn triggers the Indexer?

Did I miss anything else? Are the ApplicationEvents used for anything else?

Reply:
The only way to trigger indexing is indeed an ApplicationEvent (IndexEvent). This happens at application start via SEED variable content, at observed changes or via API (not sure if that still works).